pkg/kvstore/allocator.go

```diff
@@ -0,0 +1,539 @@
+
+	if deleteInvalid {
+		Delete(key)
+	}
+}
+
+func (a *Allocator) keyToID(key string, deleteInvalid bool) ID {
+	if !strings.HasPrefix(key, a.idPrefix) {
+		invalidKey(key, a.idPrefix, deleteInvalid)
+		return 0
+	}
+
+	id, _ := strconv.ParseUint(key[len(a.idPrefix):], 10, 64)
+	return ID(id)
+}
+
+// Naive ID allocation mechanism.
+// FIXME: This should consider a random ID in the future
+func (a *Allocator) selectAvailableID() (ID, string) {
+	for id := a.min; id <= a.max; id++ {
+		if _, ok := a.cache[id]; !ok {
+			return id, id.String()
+		}
+	}
+
+	return 0, ""
+}
+
+// RefillCache invalidates and refills the local cache. Afterwards,
+// Allocator.Cache will contain entries for all IDs to key mappings as found in
+// the KVstore.
+func (a *Allocator) RefillCache() error {
+	// fetch all /id/ keys
+	ids, err := ListPrefix(a.idPrefix)
+	if err != nil {
+		return err
+	}
+
+	newCache := IDMap{}
+	for k, v := range ids {
+		if id := a.keyToID(k, true); id != 0 {
+			key, err := a.keyType.PutKey(string(v))
+			if err == nil {
+				newCache[id] = key
+			}
+		}
+	}
+
+	a.mutex.Lock()
+	a.cache = newCache
+	if !a.disableCache {
+		a.validCache = true
+	}
+	a.mutex.Unlock()
+
+	return nil
+}
+
+func (a *Allocator) allocate(key AllocatorKey, k []byte) (ID, error) {
+	a.mutex.RLock()
+	if !a.validCache {
+		// unlock for the duration of the cache refresh
+		a.mutex.RUnlock()
+
+		if err := a.RefillCache(); err != nil {
+			return 0, err
+		}
+
+		a.mutex.RLock()
+	}
+
+	id, strID := a.selectAvailableID()
+	a.mutex.RUnlock()
+	if id == 0 {
+		return 0, fmt.Errorf("no more identifiers available")
+	}
+
+	trace("selected available key %d", id)
+
+	// create /id/<ID> and fail if it already exists
+	err := CreateOnly(a.idPrefix+strID, k, false)
+	if err == nil {
+		// Add it to local cache
+		a.mutex.Lock()
+		a.cache[id] = key
+		a.mutex.Unlock()
+
+		// return allocated ID as JSON
+		return id, nil
+	}
+
+	// Creation failed. Another agent most likely beat us to allocting this
+	// ID. Retry and also mark cache as invalid
+	a.mutex.Lock()
+	a.validCache = false
+	a.mutex.Unlock()
+
+	return 0, fmt.Errorf("CAS operation failed: %s", err)
+}
+
+func (a *Allocator) createValueNodeKey(key string, newID ID) error {
+	newIDjson, err := json.Marshal(newID)
+	if err != nil {
+		return fmt.Errorf("unable to marshal value: %s", err)
+	}
+
+	// add a new key /value/<key>/<node> to account for the reference
+	// The key is protected with a TTL/lease and will expire after LeaseTTL
+	masterKey := a.idPrefix + newID.String()
+	valueKey := a.valuePrefix + key + "/" + a.suffix
+	if err := CreateIfExists(masterKey, valueKey, newIDjson, true); err != nil {
+		return err
+	}
+
+	a.localKeys.allocate(key, newID)
+
+	return nil
+}
+
+// AllocatorKey is the interface to implement in order for a type to be used as
+// key for the allocator
+type AllocatorKey interface {
+	GetKey() string
```

Comment:
Please add a comment for every method here. I don't understand how to use those.